docs/discounts-lists/index.md

```diff
@@ -0,0 +1,496 @@
+
+# `List.nth_opt` is unsafe
+
+The problem with using `List.nth_opt` is that it can still raise an
+`Invalid_argument` exception if the value of `n` is negative. You can confirm
+this by hovering over `nth_opt` and reading the popup or [read its
+documentation](https://melange.re/v1.0.0/api/re/melange/Stdlib/List/#val-nth_opt).
+While this is unlikely to cause a problem inside `Discount.getFreeBurger`, it's
+best to avoid unsafe functions except in special circumstances, for example if a
+function needs to be as fast as possible. Of course, it's possible to implement
+a completely safe version of `nth_opt`, which is an exercise at the end of this
+chapter.
+
+---
+
+Mazel tov! You've implemented the burger discounts in a way that is more
+maintainable, and you've also learned a lot about lists along the way. In the
+next chapter, we'll finally use the discount logic to reduce the final price of
+an order.
+
+## Overview
+
+- Lists are immutable
+- You can pattern match on a whole list, even if you don't know its length
+- The `List` module contains most of the functions you'll need for
+  dealing with lists
+- The runtime representation of lists:
+  - Empty list → `0`
+  - Nonempty list → a JavaScript object with the fields `hd` (for head) and `tl`
+    (for tail)
+- The delimiters for list literals are `[]`
+- You can use list spread syntax (`...`) to add elements to the front of a list
+  or pattern match on the tail of a list
+  - The names of equivalent functions might not match the names in `Js.Array`
+- Documentation comments show up in editor hover popups and generated
+  documentation pages
+  - They can be attached to functions, modules, types, and
+  variables
+- The `ListLabels` module contains the same functions as in `List`, but they
+  have labeled arguments instead of positional arguments
+- The placeholder operator (`_`) can be used to override the position of the
+  piped argument when using the pipe last operator
+
+## Exercises
+
+<b>1.</b> There are a couple ways to improve `Discount.getFreeBurger`:
+
+- Instead of `StdLib.compare`, use a type-specific compare function. Examples
+  of a type-specific compare functions are `Bool.compare` or `String.compare`.
+- Use
+  [List.filter_map](https://melange.re/v2.2.0/api/re/melange/Stdlib/List/#val-filter_map)
+
+::: details Solution
+
+<<< Discount.re#get-free-burger-improved
+
+Prefer type-specific compare functions like [Float.compare](https://melange.re/v1.0.0/api/re/melange/Stdlib/Float/#val-compare) over polymorphic
+`compare`.
+
+:::
+
+<b>2.</b> Add a safe `Lst.nth` function which returns the nth element of a list
+encased in `Some`. If the nth element doesn’t exist, return `None`. Make sure
+the function can be used with the pipe last operator without the use of the
+placeholder operator. Refactor `Discount.getFreeBurger` to use your new
+function.
+
+::: tip
+
+Note that `Lst` is not a typo. If we name our new module `List`, it will shadow
```

Comment:
Maybe call it `EnhancedList`? Or `SafeList`? That way the tip can be skipped.

Reply:
Renamed in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/e9a4db32b0d9354f8a4455b60a7723f3b113767a